### ISSUE TITLE
Can still read cartridge data after removing cartridge

### ISSUE THREAD

**Issue by lxylxy123456** [UCDClassNitta/riscv-console#12]
Currently git version is `39a982b0d0237a36a130501534bd2f9c2ff8e2e1`

To reproduce:
1. Replace main of `riscv-example` with the file at the end
2. Compile `riscv-example`, also compile a cartridge image
3. Start the simulator, load `riscv-example`. Now the simulator should show 5 lines of `00000000`.
4. INS the cartridge, and the simulator should show 5 lines of cartridge data.
5. Press REM, now the simulator still shows cartridge data. However, 5 lines of `00000000` are expected.

Also, I guess there maybe another possible bug due to a strange error of my program's execution:
* Suppose cartridge A's content is `aaaaaa00`
* Suppose cartridge B's content is `bbbb0000`
* Then after loading A, and replacing it with B, the content in the simulator is effectively `bbbbaa00` when executed
* However, I have not verified it yet

`riscv-example/src/main.c`:
```
#include <stdint.h>

volatile int global = 42;
volatile uint32_t controller_status = 0;


volatile uint8_t *video_text_data_get(int x, int y) {
	return (volatile uint8_t *)(0x500FE800 + x + y * 64);
}

void print_hex(int x, int y, uint8_t n) {
	// Print a 0-15 number as hex value, '?' when error
	if (n < 10)
		*video_text_data_get(x, y) = n + '0';
	else if (n < 16)
		*video_text_data_get(x, y) = n - 10 + 'a';
	else
		*video_text_data_get(x, y) = '?';
}

void print_uint32(int x, int y, uint32_t n) {
	// Print uint32_t as hex value
	// Right is lower memory address
	for (int i = 0; i < 8; i++)
		print_hex(x + i, y, (n >> (32 - 4 - i * 4)) & 0xf);
}

volatile char *VIDEO_MEMORY = (volatile char *)(0x50000000 + 0xFE800);
int main() {
    int a = 4;
    int b = 12;
    int last_global = 42;
    int x_pos = 12;

    VIDEO_MEMORY[0] = 'H';
    VIDEO_MEMORY[1] = 'e';
    VIDEO_MEMORY[2] = 'l';
    VIDEO_MEMORY[3] = 'l';
    VIDEO_MEMORY[4] = 'o';
    VIDEO_MEMORY[5] = ' ';
    VIDEO_MEMORY[6] = 'W';
    VIDEO_MEMORY[7] = 'o';
    VIDEO_MEMORY[8] = 'r';
    VIDEO_MEMORY[9] = 'l';
    VIDEO_MEMORY[10] = 'd';
    VIDEO_MEMORY[11] = '!';
    VIDEO_MEMORY[12] = 'X';


    while (1) {
        int c = a + b + global;
        if(global != last_global){
        	print_uint32(0, 2, ((volatile uint32_t *)0x20000000)[0]);
        	print_uint32(0, 3, ((volatile uint32_t *)0x20000000)[1]);
        	print_uint32(0, 4, ((volatile uint32_t *)0x20000000)[2]);
        	print_uint32(0, 5, ((volatile uint32_t *)0x20000000)[3]);
        	print_uint32(0, 6, ((volatile uint32_t *)0x20000000)[4]);
            if(controller_status){
                VIDEO_MEMORY[x_pos] = ' ';
                if(controller_status & 0x1){
                    if(x_pos & 0x3F){
                        x_pos--;
                    }
                }
                if(controller_status & 0x2){
                    if(x_pos >= 0x40){
                        x_pos -= 0x40;
                    }
                }
                if(controller_status & 0x4){
                    if(x_pos < 0x8C0){
                        x_pos += 0x40;
                    }
                }
                if(controller_status & 0x8){
                    if((x_pos & 0x3F) != 0x3F){
                        x_pos++;
                    }
                }
                VIDEO_MEMORY[x_pos] = 'X';
            }
            last_global = global;
        }
    }
    return 0;
}
```

**Comment by cjnitta** [comment]
This is related to https://github.com/UCDClassNitta/riscv-console/issues/13 as well. The instruction cache needs to be flushed when the cartridge is removed. I need to add the ability to flush only a portion of the cache. This should fix both problems.

**Comment by cjnitta** [comment]
Just pushed, see if it fixes the issues.

**Comment by lxylxy123456** [comment]
Neither issues are fixed on my end (git `62dad941d81528b2d6f6c16e41c7790bfe8aa55f`)

**Comment by cjnitta** [comment]
I thought the instructions weren't updating in https://github.com/UCDClassNitta/riscv-console/issues/13? They updated for me when I tested. Can you test this?
```
--- a/riscv-sim/src/RISCVConsole.cpp
+++ b/riscv-sim/src/RISCVConsole.cpp
@@ -404,6 +404,9 @@ bool CRISCVConsole::RemoveCartridge(){
     DInstructionStrings = DFirmwareInstructionStrings;
     DInstructionAddressesToIndices = DFirmwareAddressesToIndices;
     MarkBreakpointStrings();
+    DCartridgeFlash->WriteEnabled(true);
+    DCartridgeFlash->EraseAll();
+    DCartridgeFlash->WriteEnabled(false);
     DCPUCache->FlushRange(DCartridgeMemoryBase, DCartridgeMemorySize);
     DChipset->RemoveCartridge();
     if(CurrentState == to_underlying(EThreadState::Run)){
```

**Comment by lxylxy123456** [comment]
I think it also does not work (the 3 lines added at `riscv-sim/src/RISCVConsole.cpp` line 407)

I made another example, which explains why my program runs weird.
First clone branch `bug6` of my clone: https://github.com/lxylxy123456/riscv-console/tree/bug6
(coming soon)

**Comment by cjnitta** [comment]
I tested the above with the patch that has been committed, and it seems to replace all of the memory locations with 0. Also the instructions are updated so https://github.com/UCDClassNitta/riscv-console/issues/13 seems fixed as well. I can't clone bug6, I'm getting that the repository is not found.

**Comment by lxylxy123456** [comment]
Sorry for being confusing. That URL is for a web browser. The repo for `git` is git@github.com:lxylxy123456/riscv-console , and branch name is `bug6`.

Currently according to my testing, it looks like #12 is fixed, but #13 and `bug6` are partially fixed.
When a cartridge is already inserted, the simulator works fine if the user press "REM" and then press "INS" and then select a file. But if the user just press "INS" and selects a file, the bug exists.

**Comment by lxylxy123456** [comment]
Here is a video

https://user-images.githubusercontent.com/9337403/107126037-41812b00-6862-11eb-94c2-820572da83a5.mp4

**Comment by cjnitta** [comment]
I had the firmware flushing instead of the cartridge on the insert. Try it now. I tested the bug6, and I think it works as you were stating it should.

**Comment by lxylxy123456** [comment]
Looks like the issues are fixed (both #12 and #13). Please feel free to close this issue.
